kotest-common/src/commonMain/kotlin/io/kotest/common/testTimeSource.kt

```diff
@@ -0,0 +1,50 @@
+package io.kotest.common
+
+import kotlin.coroutines.ContinuationInterceptor
+import kotlin.coroutines.coroutineContext
+import kotlin.time.TimeSource
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.test.TestDispatcher
+import kotlinx.coroutines.test.TestCoroutineScheduler
+
+/**
+ * Returns the [TimeSource] used in tests.
+ *
+ * This is [TimeSource.Monotonic] or virtual time, depending on the scheduler in use.
+ */
+@OptIn(KotestInternal::class)
+suspend fun testTimeSource(): TimeSource =
+   coroutineContext.testCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic
+
+@KotestInternal
+val CoroutineContext.testCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = when (val dispatcher = this[ContinuationInterceptor]) {
+      is TestDispatcher -> dispatcher.scheduler
+      else -> null
+   }
+
+/**
+ * Returns the [TimeSource] used for non-deterministic test functions.
+ *
+ * This is [TimeSource.Monotonic] or – if separately enabled – virtual time, depending on the scheduler in use.
+ */
+@OptIn(KotestInternal::class)
+suspend fun nonDeterministicTestTimeSource(): TimeSource =
+   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic
+
+@OptIn(KotestInternal::class)
+private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
+   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```

Comment:
I think it would be better if these functions were also marked as `@KotestInternal`.

- Enabling them requires using the ``, which is also marked as `@KotestInternal`, so it makes sense to me to also mark the utils as internals.
- They're specifically for Kotest internal testing, and so shouldn't be confused with the regular `testTimeSource()` that most users should be using.

```suggestion
@KotestInternal
suspend fun nonDeterministicTestTimeSource(): TimeSource =
   coroutineContext.nonDeterministicTestCoroutineSchedulerOrNull?.timeSource ?: TimeSource.Monotonic

@KotestInternal
private val CoroutineContext.nonDeterministicTestCoroutineSchedulerOrNull: TestCoroutineScheduler?
   get() = if (this[NonDeterministicTestVirtualTimeEnabled] != null) testCoroutineSchedulerOrNull else null
```